Sources/HummingbirdCore/Server/HTTP/HTTPChannelSetup.swift

```diff
@@ -0,0 +1,114 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Hummingbird server framework project
+//
+// Copyright (c) 2023 the Hummingbird authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See hummingbird/CONTRIBUTORS.txt for the list of Hummingbird authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import NIOCore
+import NIOHTTP1
+
+/// Protocol for HTTP channels
+public protocol HTTPChannelSetup: HBChannelSetup where In == HTTPServerRequestPart, Out == SendableHTTPServerResponsePart {
+    var responder: @Sendable (HBHTTPRequest, Channel) async throws -> HBHTTPResponse { get }
+}
+
+/// Internal error thrown when an unexpected HTTP part is received eg we didn't receive
+/// a head part when we expected one
+enum HTTPChannelError: Error {
+    case unexpectedHTTPPart(HTTPServerRequestPart)
+    case closeConnection
+}
+
+extension HTTPChannelSetup {
+    public func handle(asyncChannel: NIOAsyncChannel<In, Out>, logger: Logger) async {
+        do {
+            try await withThrowingTaskGroup(of: Void.self) { group in
+                let responseWriter = HBHTTPServerBodyWriter(outbound: asyncChannel.outbound)
+                var iterator = asyncChannel.inbound.makeAsyncIterator()
```

Comment:
THIS is absolutely amazing. THAT's why I couldn't get my implementation to work as well as it should